File not found errors for config files now indicate which file is missing

kadinsayani opened the pull request:
Closes #25.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced error handling for configuration files, including specific error messages for missing main configuration files, included configuration files, and plugin files.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

kadinsayani commented:
@coderabbitai full review

kadinsayani commented:
@coderabbitai full review

sporkmonger commented on crates/config/src/errors.rs:
```diff
@@ -3,6 +3,12 @@
 pub enum ConfigFileError {
     #[error(transparent)]
     IO(#[from] std::io::Error),
+    #[error("no such file or directory: {0}")]
+    NotFound(String),
+    #[error("no such file or directory: {0}")]
+    IncludedFileNotFound(String),
```
I think `NotFound` should be `ConfigNotFound` and should have a differentiated error message from the current `IncludedFileNotFound`. Additionally, I'd probably rename `IncludedFileNotFound` to `IncludedConfigNotFound`.

I think these should be the error messages:
1. `"config file not found: {0}"`
2. `"included config file not found: {0}"`

sporkmonger commented on crates/config/src/errors.rs:
```diff
@@ -3,6 +3,12 @@
+    NotFound(String),
+    #[error("no such file or directory: {0}")]
+    IncludedFileNotFound(String),
+    #[error("plugin .wasm file not found: {0}")]
```
Probably drop the `.wasm` here. Technically I don't think we prevent files from using a different extension and we might already even support loading `.wat` files instead because the underlying `wasmtime` dependency does. On the off-chance someone takes advantage of that for some reason, the error message could be confusing.

sporkmonger commented on crates/config/src/toml.rs:
```diff
@@ -1111,7 +1127,7 @@ mod tests {
             .unwrap_err()
             .to_string()
-            .starts_with("No such file or directory"));
+            .starts_with("no such file or directory"));
```
We probably want to additionally have the test confirm the filename is present now, but for the second assertion, use a `contains` check to avoid having to be concerned with whether the path in the error message is relative or absolute.

kadinsayani commented:
@sporkmonger, I see that you self assigned this issue, would you still like me to resolve the requested changes?

sporkmonger commented:
Sorry for the confusion, meant to add myself as a reviewer, not as the assignee.

sporkmonger commented on crates/config/src/errors.rs:
```diff
@@ -3,6 +3,12 @@
+    #[error("no such file or directory: {0}")]
+    IncludedFileNotFound(String),
+    #[error("plugin .wasm file not found: {0}")]
+    PluginWasmNotFound(String),
```
@coderabbitai Yes, please draft similar test cases that cover `IncludedFileNotFound` and `PluginWasmNotFound` error scenarios.

sporkmonger commented on crates/config/src/toml.rs:
```diff
@@ -524,7 +524,17 @@ where
             loaded_files.insert(path_string);
         }
-        let toml_data = fs::read_to_string(config_path)?;
+        let toml_data = match fs::read_to_string(config_path) {
+            Ok(data) => data,
+            Err(err) => match err.kind() {
+                std::io::ErrorKind::NotFound => {
+                    return Err(ConfigFileError::ConfigNotFound(
+                        config_path.as_ref().to_string_lossy().into_owned(),
+                    ));
+                }
+                _ => Err(ConfigFileError::IO(err)),
+            }?,
+        };
```
This is a decent example of a case where good old `println!` debugging is useful.

I added this just before the assertions:
```rust
println!("{}", err.to_string());
```

Output:
```txt
---- toml::tests::test_load_config_missing_include stdout ----
included config file not found: 'tests/does_not_exist.toml'
thread 'toml::tests::test_load_config_missing_include' panicked at crates/config/src/toml.rs:1128:9:
assertion failed: err.to_string().starts_with("config file not found")
```

The assertions you've written check for the entire config file being missing. That's the first error scenario. The test is configured to create the second include file missing scenario. There should be all 3 test scenarios present.

sporkmonger replied:
You may need to create a new `.toml` file to support the missing plugin file test. Probably rename `missing.toml` as well to clarify that it's a missing plugin reference rather than a missing plugin file, because that will now be ambiguous.

sporkmonger commented on crates/config/src/toml.rs:
```diff
@@ -1108,10 +1123,9 @@ mod tests {
-            .to_string()
-            .starts_with("No such file or directory"));
+        let err = result.unwrap_err();
+        assert!(err.to_string().starts_with("config file not found"));
+        assert!(err.to_string().contains("missing_include.toml"));
```
```suggestion
        assert!(err.to_string().starts_with("included config file not found"));
        assert!(err.to_string().contains("tests/does_not_exist.toml"));
```

sporkmonger reviewed (changes requested):
Here's the fix for the failing test, and then we'd need the other two tests to be added.

kadinsayani commented:
I've added the missing config file test case, and a function stub for the missing plugin file test case. I've also fixed the failing test case. 

The only outstanding item is to return the `PluginNotFound` error when a missing plugin file is referenced. When you have a chance, can you please advise on when this error should be returned? Should it be returned by `resolve_plugins` which currently returns `ResolutionError::Missing`? Thanks!

sporkmonger commented on crates/config/tests/missing_plugin.toml:
```diff
@@ -0,0 +1,4 @@
+[[plugin]]
+ref = "blank_slate"
+path = "bulwark_blank_slate.wasm"
```
I don't personally do much TDD, but this is a good use-case for it. I don't know where there right place in the code to return this error is, off the top of my head, but I know what test case should trigger it.

```suggestion
ref = "does_not_exist"
path = "does_not_exist.wasm"
```

You'll also need to update this test file with a resource that points to this plugin. Something like:

```toml
[[resource]]
routes = ["/"]
plugins = ["does_not_exist"]
timeout = 25
```

The `test_load_config_missing_plugin` should attempt to load this file. It should error out, but initially won't have the correct error message. Before you write the code, you can write an assertion that says it should contain `"included plugin file not found"`. This would be expected to fail, but you can run the test with backtraces enabled and I think that should show you exactly where the existing error needs to be changed to use the new variant.

sporkmonger reviewed (changes requested):
Making good progress, thanks for your work on this!

sporkmonger commented on crates/config/src/toml.rs:
```diff
@@ -1143,7 +1157,17 @@ mod tests {
+        let result = load_config("tests/missing_plugin.toml");
+        assert!(result.is_err());
+        let err = result.unwrap_err();
+        println!("{}", err);
```
You've left a stray debug statement in. This should be removed.

sporkmonger commented on crates/config/src/toml.rs:
```diff
@@ -1102,16 +1131,43 @@ mod tests {
-            .starts_with("No such file or directory"));
+            .starts_with("included plugin file not found"));
+        assert!(err.to_string().contains("does_not_exist.wasm"));
+        println!("{}", err);
```
Same here.

sporkmonger reviewed (changes requested):
Just needs the debug cleanup and then I think this is all set!

sporkmonger reviewed (approved):
Looks good to me! Thank you!
